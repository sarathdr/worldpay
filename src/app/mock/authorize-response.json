{
  "outcome": "authorized",
  "issuer": {
    "authorizationCode": "0"
  },
  "_links": {
    "payments:cancel": {
      "href": "https://try.access.worldpay.com/payments/authorizations/cancellations/eyJrIjoiazNhYjYzMiJ9"
    },
    "payments:settle": {
      "href": "https://try.access.worldpay.com/payments/settlements/full/eyJrIjoiazNhYjYzMiJ9"
    },
    "payments:partialSettle": {
      "href": "https://try.access.worldpay.com/payments/settlements/partials/eyJrIjoiazNhYjYzMiJ9"
    },
    "payments:events": {
      "href": "https://try.access.worldpay.com/payments/events/eyJrIjoiazNhYjYzMiJ9"
    },
    "curies": [{
      "name": "payments",
      "href": "https://try.access.worldpay.com/rels/payments/{rel}",
      "templated": true
    }]
  }
}
