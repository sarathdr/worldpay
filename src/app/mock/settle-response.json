{
  "_links": {
    "payments:refund": {
      "href": "https://try.access.worldpay.com/payments/settlements/refunds/full/eyJrIjoiazNhYjYzMiJ9"
    },
    "payments:partialRefund": {
      "href": "https://try.access.worldpay.com/payments/settlements/refunds/partials/eyJrIjoiazNhYjYzMiJ9"
    },
    "payments:events": {
      "href": "https://try.access.worldpay.com/payments/events/eyJrIjoiazNhYjYzMiJ9"
    },
    "curies": [
      {
        "name": "payments",
        "href": "https://try.access.worldpay.com/rels/payments/{rel}",
        "templated": true
      }
    ]
  }
}
