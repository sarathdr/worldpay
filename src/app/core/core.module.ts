import { APP_INITIALIZER, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ConfigService } from './services/config.service';
import { PaymentService } from './services/payment.service';
import { CartService } from './services/cart.service';
import { HttpClientModule } from '@angular/common/http';
import { FormsModule } from '@angular/forms';


export const fetchConfig = (appConfig: ConfigService) => {
  return () => appConfig.getPaymentUrls();
};

@NgModule({
  imports: [
    CommonModule,
    FormsModule
  ],
  providers: [
    ConfigService,
    PaymentService,
    HttpClientModule,
    CartService,
    {
      provide: APP_INITIALIZER,
      useFactory: fetchConfig,
      deps: [ConfigService],
      multi: true
    }
  ]
})
export class CoreModule { }
