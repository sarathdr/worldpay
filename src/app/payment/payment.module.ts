import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AuthorizeComponent } from './authorize/authorize.component';
import { AuthorizeContainer } from './authorize/authorize.container';
import { ConfirmComponent } from './confirm/confirm.component';
import { ConfirmContainer } from './confirm/confirm.container';
import { RefundComponent } from './refund/refund.component';
import { RefundContainer } from './refund/refund.container';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { MatButtonModule, MatFormFieldModule, MatInputModule, MatSelectModule } from '@angular/material';
import { AuthorizeFormComponent } from './authorize/authorize-form.component';
import { ConfirmPaymentGuard } from './guards/confirm-payment.guard';
import { ExpiryDateModule } from '../core/directives/expiry-date.module';
import { RefundPaymentGuard } from './guards/refund-payment.guard';
import { RefundSuccessComponent } from './refund/refund-success.component';

@NgModule({
  declarations: [
    AuthorizeComponent,
    AuthorizeFormComponent,
    AuthorizeContainer,
    ConfirmComponent,
    ConfirmContainer,
    RefundComponent,
    RefundSuccessComponent,
    RefundContainer
  ],
  imports: [
    CommonModule,
    FormsModule,
    RouterModule.forChild([
      {
        path: '',
        component: AuthorizeContainer
      },
      {
        path: 'confirm',
        component: ConfirmContainer,
        canActivate:[ConfirmPaymentGuard],
        canDeactivate:[ConfirmPaymentGuard]
      },
      {
        path: 'refund',
        component: RefundContainer,
        canActivate:[RefundPaymentGuard],
        canDeactivate:[RefundPaymentGuard]
      }
    ]),
    MatFormFieldModule,
    MatInputModule,
    MatSelectModule,
    ReactiveFormsModule,
    MatButtonModule,
    ExpiryDateModule
  ],
  providers:[
    ConfirmPaymentGuard,
    RefundPaymentGuard
  ]
})
export class PaymentModule { }
