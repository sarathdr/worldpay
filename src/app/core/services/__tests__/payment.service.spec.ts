import { TestBed } from '@angular/core/testing';
import { HttpClient } from '@angular/common/http';
import { cold } from 'jasmine-marbles';
import { ConfigService } from '../config.service';
import { PaymentService } from '../payment.service';
import * as authorizeResponse from '../../../mock/authorize-response.json';
import { AuthorizeRequest } from '../../models/payment.interface';
import * as settleResponse from '../../../mock/settle-response.json';

describe('PaymentService', () => {

  let service: PaymentService;
  let post: jest.Mock;

  const getAuthorizePaymentUrl = jest.fn();

  beforeEach(() => {
    post = jest.fn();
    getAuthorizePaymentUrl.mockReturnValue('http://mockurl');

    TestBed.configureTestingModule({
      providers: [
        PaymentService,
        {
          provide: HttpClient,
          useValue: {
            post
          },
        },
        {
          provide: ConfigService,
          useValue: {
            getAuthorizePaymentUrl
          }
        }
      ]
    });

    service = TestBed.get(PaymentService);

  });

  describe('authorizePayment ', () => {

    it('should authorize, save urls and reset', () => {
      const response$ = cold('-a', { a: authorizeResponse });
      const expected$ = cold('-b', { b: authorizeResponse });
      const subs = '^';

      post.mockReturnValue(response$);

      expect(service.authorizePayment({} as AuthorizeRequest)).toBeObservable(expected$);
      expect(response$).toHaveSubscriptions(subs);
      expect(service.hasAuthorized()).toBeTruthy();
      expect(service.getCancelUrl()).toMatchSnapshot();
      expect(service.getSettleUrl()).toMatchSnapshot();

      service.resetAuthorize();
      expect(service.hasAuthorized()).toBeFalsy();
    });

    it('should handle error', () => {
      const expected$ = cold('-(a|)', { a: { status: 400 } });
      const response$ = cold('-#', {}, { status: 400 });
      post.mockReturnValue(response$);

      expect(service.authorizePayment({} as AuthorizeRequest)).toBeObservable(expected$);
    });

  });


  describe('confirmPayment ', () => {

    beforeEach(() => {
      const mockSettleUrl = jest.fn();
      service.getSettleUrl = mockSettleUrl;
      mockSettleUrl.mockReturnValue('http://mock-settle');
    });

    it('should confirm, save and reset urls', () => {

      const response$ = cold('-a', { a: settleResponse });
      const expected$ = cold('-b', { b: settleResponse });
      const subs = '^';

      post.mockReturnValue(response$);

      expect(service.confirmPayment()).toBeObservable(expected$);
      expect(response$).toHaveSubscriptions(subs);
      expect(service.hasSettled()).toBeTruthy();
      expect(service.getRefundUrl()).toMatchSnapshot();

      service.resetRefund();
      expect(service.hasSettled()).toBeFalsy();
    });

    it('should handle error', () => {
      const response$ = cold('-#', {}, { status: 400 });
      const expected$ = cold('-(a|)', { a: { status: 400 } });
      post.mockReturnValue(response$);

      expect(service.confirmPayment()).toBeObservable(expected$);
    });

  });

  describe('cancelPayment ', () => {

    beforeEach(() => {
      const mockGetCancelUrl = jest.fn();
      service.getCancelUrl = mockGetCancelUrl;
      mockGetCancelUrl.mockReturnValue('http://mock-cancel');
    });

    it('should cancel payment', () => {
      const response$ = cold('-a', { a: '*' });
      const expected$ = cold('-b', { b: '*' });
      const subs = '^';

      post.mockReturnValue(response$);
      expect(service.cancelPayment()).toBeObservable(expected$);
      expect(response$).toHaveSubscriptions(subs);
    });

    it('should handle error', () => {
      const expected$ = cold('-(a|)', { a: { status: 400 } });
      const response$ = cold('-#', {}, { status: 400 });
      post.mockReturnValue(response$);

      expect(service.cancelPayment()).toBeObservable(expected$);
    });

  });

  describe('refundPayment ', () => {

    beforeEach(() => {
      const mockGetRefundUrl = jest.fn();
      service.getRefundUrl = mockGetRefundUrl;
      mockGetRefundUrl.mockReturnValue('http://mock-refund');
    });

    it('should refund payment', () => {
      const response$ = cold('-a', { a: '*' });
      const expected$ = cold('-b', { b: '*' });
      const subs = '^';

      post.mockReturnValue(response$);

      expect(service.refundPayment()).toBeObservable(expected$);
      expect(response$).toHaveSubscriptions(subs);
    });

    it('should handle error', () => {
      const expected$ = cold('-(a|)', { a: { status: 400 } });
      const response$ = cold('-#', {}, { status: 400 });
      post.mockReturnValue(response$);

      expect(service.refundPayment()).toBeObservable(expected$);
    });

  });

});
