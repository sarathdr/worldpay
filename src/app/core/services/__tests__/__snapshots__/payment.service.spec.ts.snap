// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentService authorizePayment  should authorize, save urls and reset 1`] = `"https://try.access.worldpay.com/payments/authorizations/cancellations/eyJrIjoiazNhYjYzMiJ9"`;

exports[`PaymentService authorizePayment  should authorize, save urls and reset 2`] = `"https://try.access.worldpay.com/payments/settlements/full/eyJrIjoiazNhYjYzMiJ9"`;

exports[`PaymentService confirmPayment  should confirm, save and reset urls 1`] = `"https://try.access.worldpay.com/payments/settlements/refunds/full/eyJrIjoiazNhYjYzMiJ9"`;
